File: solve/lua/insert.lua
require "common" 

skeleton(function ()
    for i = 2, n do
        local temp = a[i]
        local j = i - 1
        repeat
            if a[j] <= temp then break end
            a[j + 1] = a[j]
            j = j - 1
        until j == 0
        a[j + 1] = temp
    end     
end)

File: solve/cpp/insert.cc
#include "common.hh"

void insertsort()
{
    for (int *p = a + 1; p != a + n; ++p) 
    {
        int temp = *p; 
        int *q;
        for (q = p - 1; q >= a; --q)
        {
            if (*q > temp)
            {
                *(q + 1) = *q;
            }
            else
            {
                break;
            }
        }
        *(q + 1) = temp;
    }
}

int main(int argc, char *argv[])
{
    if (init(argc, argv) == 0)
    {
        insertsort();
        output();
        return 0;
    } 
    return 1;
}
